design-documents/graph-ql/coverage/storefront-route.md

```diff
@@ -0,0 +1,292 @@
+```
+
+Any type which can be a page model must implement `RoutableInterface` in addition to any other interfaces.
+
+```graphql
+type CmsPage implements RoutableInterface {
+  content: String
+  content_heading: String
+  meta_description: String
+  meta_keywords: String
+  meta_title: String
+  page_layout: String
+  title: String
+  url_key: String
+  display_metadata: String
+}
+
+# multiple interfaces
+type SimpleProduct implements ProductInterface & RoutableInterface {
+  # [...other property definitions]
+  display_metadata: String
+}
+```
+
+### Route Query
+
+To enable clients to query based on route, add a root query to the schema.
+
+```graphql
+type Query {
+    route(url: String!): RoutableInterface
+}
+```
+
+This query would replace `urlResolver` in the PWA implementation; `urlResolver` could then be deprecated.
+
+The simplest route query would be:
+
+```graphql
+route(url: "/product-foo.html") {
+  __typename
+}
+```
+
+The response to such a query would be:
+
+```json
+{
+  "data": {
+    "route": {
+      "__typename": "SimpleProduct"
+    }
+  }
+}
+```
+
+This query alone would not be useful without placing a subsequent query for entity data.
+Since `RoutableInterface` entities are also page models, a more useful query would contain a typed fragment for supported components:
+
+```graphql
+route(url: "/product-foo.html") {
+  __typename
+  display_metadata
+  ...on SimpleProduct {
+    name
+    description
+    media_gallery_entries {
+      url
+    }
+  }
+  ...on CmsPage {
+    title
+    content
+  }
+}
+```
+
+The response would be:
+
+```json
+{
+  "data": {
+    "route": {
+      "__typename": "SimpleProduct",
+      "display_metadata": "featured",
+      "name": "Product Foo",
+      "description": "Example product",
+      "media_gallery_entries": [
+        {
+          "url": "/media/catalog/product/foo.jpg"
+        }
+      ]
+    }
+  }
+}
+```
+
+This combines entity type, entity data for two types of concrete entity, and display metadata in a single query.
+The PWA uses `__typename` to route to the correct display component, `display_metadata` to optionally use an labeled alternate view (e.g. `featured`, `dark` or `microsite_A`), and the rest of the properties as content to render.
+
+## Use Cases
+
+The `RoutableInterface` covers all the use cases for PWA Studio to support core Magento functionality and familiar Magento extensibility.
+With `RoutableInterface`, a Magento instance describes its sitemap in its GraphQL schema.
+Extension developers create custom pages by adding an implementation of `RoutableInterface` to the store schema.
+
+### SEO URL for catalog and CMS pages
+
+The above examples demonstrate this idea. A full-featured PWA would ideally have typed fragments for each supported page type. When the page type is unfamiliar to the PWA, the PWA can still handle it by rendering a general purpose "app shell" and displaying a 404 or other custom page.
+
+```graphql
+route(url: "/product-foo.html") {
```

Comment:
Because the RoutableInterface concept also provides:
- Build-time analysis to determine which entity types will need to be routable and build an efficient minimal application
- Usage of GraphQL introspection instead of magic strings
- `display_metadata` to help the PWA choose an alternate page treatment for a particular entity